File: src/main/java/io/scriptor/impl/FPRFile64.java
package io.scriptor.impl;

import io.scriptor.machine.FPRFile;
import org.jetbrains.annotations.NotNull;

import java.io.PrintStream;
import java.util.Arrays;

public final class FPRFile64 implements FPRFile {

    private final double[] values = new double[32];

    @Override
    public void reset() {
        Arrays.fill(values, 0.0);
    }

    @Override
    public void dump(final @NotNull PrintStream out) {
        for (int i = 0; i < values.length; ++i) {
            out.printf("f%-2d: %-16f  ", i, values[i]);

            if ((i + 1) % 4 == 0) {
                out.println();
            }
        }
    }

    @Override
    public float getf(final int reg) {
        if (reg == 0) {
            return 0.0f;
        }
        return (float) values[reg];
    }

    @Override
    public double getd(final int reg) {
        if (reg == 0) {
            return 0.0;
        }
        return values[reg];
    }

    @Override
    public void putf(final int reg, final float val) {
        if (reg == 0) {
            return;
        }
        values[reg] = val;
    }

    @Override
    public void putd(final int reg, final double val) {
        if (reg == 0) {
            return;
        }
        values[reg] = val;
    }

    @Override
    public int getfr(final int reg) {
        return Float.floatToRawIntBits(getf(reg));
    }

    @Override
    public long getdr(final int reg) {
        return Double.doubleToRawLongBits(getd(reg));
    }

    @Override
    public void putfr(final int reg, final int val) {
        putf(reg, Float.intBitsToFloat(val));
    }

    @Override
    public void putdr(final int reg, final long val) {
        putd(reg, Double.longBitsToDouble(val));
    }
}

File: src/main/java/io/scriptor/impl/v64/FPRFile64.kt
package io.scriptor.impl.v64

import io.scriptor.fdt.BuilderContext
import io.scriptor.fdt.NodeBuilder
import io.scriptor.machine.Device
import io.scriptor.machine.FPRFile
import io.scriptor.util.Log.format
import java.io.PrintStream

class FPRFile64 : FPRFile {

    override val machine
        get() = hart.machine

    private val hart: Hart64

    @OptIn(ExperimentalUnsignedTypes::class)
    private val values = ULongArray(32)

    constructor(hart: Hart64) {
        this.hart = hart
    }

    override fun build(context: BuilderContext<Device>, builder: NodeBuilder) {
        TODO()
    }

    @OptIn(ExperimentalUnsignedTypes::class)
    override fun dump(out: PrintStream) {
        for (i in values.indices) {
            out.print(format("f%-2d: %016x ", i, values[i]))

            if ((i + 1) % 4 == 0) {
                out.println()
            }
        }
    }

    @OptIn(ExperimentalUnsignedTypes::class)
    override fun reset() {
        values.fill(0UL)
    }

    @OptIn(ExperimentalUnsignedTypes::class)
    override fun getfr(reg: UInt): UInt = if (reg == 0U) 0U else values[reg.toInt()].toUInt()

    @OptIn(ExperimentalUnsignedTypes::class)
    override fun getdr(reg: UInt): ULong = if (reg == 0U) 0UL else values[reg.toInt()]

    @OptIn(ExperimentalUnsignedTypes::class)
    override fun put(reg: UInt, value: UInt) {
        if (reg != 0U) values[reg.toInt()] = value.toULong()
    }

    @OptIn(ExperimentalUnsignedTypes::class)
    override fun put(reg: UInt, value: ULong) {
        if (reg != 0U) values[reg.toInt()] = value
    }
}
